src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,240 @@
+                        // reference to the target
+                        return null;
+                    }
+
+                    checked
+                    {
+                        referenceCount.Value++;
+                    }
+
+                    // Must return a new instance, in order for the Dispose operation on each individual instance to
+                    // be idempotent.
+                    return new ReferenceCountedDisposable<T>(target, referenceCount);
+                }
+            }
+
+            /// <summary>
+            /// Releases the current reference, causing the underlying object to be disposed if this was the last
+            /// reference.
+            /// </summary>
+            /// <remarks>
+            /// <para>After this instance is disposed, the <see cref="TryAddReference"/> method can no longer be used to
+            /// object a new reference to the target, even if other references to the target object are still in
+            /// use.</para>
+            /// </remarks>
+            public void Dispose()
+            {
+                T instanceToDispose = null;
+                lock (_boxedReferenceCount)
+                {
+                    if (_instance == null)
+                    {
+                        // Already disposed; allow multiple without error.
+                        return;
+                    }
+
+                    _boxedReferenceCount.Value--;
+                    if (_boxedReferenceCount.Value == 0)
+                    {
+                        instanceToDispose = _instance;
+                    }
+
+                    // Ensure multiple calls to Dispose for this instance are a NOP.
+                    _instance = null;
+                }
+
+                instanceToDispose?.Dispose();
+            }
+
+            /// <summary>
+            /// Represents a weak reference to a <see cref="ReferenceCountedDisposable{T}"/> which is capable of
+            /// obtaining a new counted reference up until the point when the object is no longer accessible.
+            /// </summary>
+            public struct WeakReference
+            {
+                /// <summary>
+                /// DO NOT DISPOSE OF THE TARGET.
+                /// </summary>
+                private readonly WeakReference<T> _weakInstance;
+                private readonly StrongBox<int> _boxedReferenceCount;
+
+                public WeakReference(ReferenceCountedDisposable<T> reference)
+                    : this()
+                {
+                    if (reference == null)
+                    {
+                        throw new ArgumentNullException(nameof(reference));
+                    }
+
+                    var instance = reference._instance;
+                    var referenceCount = reference._boxedReferenceCount;
+                    if (instance == null)
+                    {
+                        // The specified reference is already not valid.
```

Comment:
No, this is intended to be the equivalent of `new WeakReference(null)`. It's allowed, but when you get the target it will return null. The special case is simply to avoid an unnecessary allocation when this happens.